agent/system_metrics.go

```diff
@@ -0,0 +1,303 @@
+
+	return nil
+}
+
+func (s *systemMetricsCollector) timedCollect(ctx context.Context, mc metricCollector, stream *metrics.SystemMetricsWriteCloser) {
+	timer := time.NewTimer(0)
+	defer func() {
+		s.catcher.Add(errors.Wrap(recovery.HandlePanicWithError(recover(), nil, fmt.Sprintf("panic in system metrics stream for id %s and metricType %s", s.id, mc.Name())), ""))
+		s.catcher.Add(errors.Wrap(stream.Close(), fmt.Sprintf("problem closing system metrics stream for id %s and metricType %s", s.id, mc.Name())))
+		timer.Stop()
+		s.stream.Done()
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
+			data, err := mc.Collect()
+			if err != nil {
+				s.catcher.Add(errors.Wrapf(err, "problem collecting system metrics data for id %s and metricType %s", s.id, mc.Name()))
+				return
+			}
+			_, err = stream.Write(data)
+			if err != nil {
+				s.catcher.Add(errors.Wrapf(err, "problem writing system metrics data to stream for id %s and metricType %s", s.id, mc.Name()))
+				return
+			}
+			_ = timer.Reset(s.interval)
+		}
+	}
+}
+
+func (s *systemMetricsCollector) closeOnCancel(outerCtx, streamingCtx context.Context) {
+	defer s.close.Done()
+	for {
+		select {
+		case <-outerCtx.Done():
+			if s.streamingCancel != nil {
+				s.streamingCancel()
+			}
+			s.cleanup()
+			grip.Error(s.catcher.Resolve())
+			return
+		case <-streamingCtx.Done():
+			s.cleanup()
+			return
+		}
+	}
+}
+
+func (s *systemMetricsCollector) cleanup() {
+	s.stream.Wait()
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.client != nil {
+		ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+		defer cancel()
+		if s.id != "" {
+			s.catcher.Add(errors.Wrap(s.client.CloseSystemMetrics(ctx, s.id, s.catcher.HasErrors()), fmt.Sprintf("error closing out system metrics object for id %s", s.id)))
+		}
+		s.catcher.Add(errors.Wrap(s.client.CloseClient(), fmt.Sprintf("error closing system metrics client for id %s", s.id)))
+	}
+	s.closed = true
+}
+
+// Close cleans up any remaining connections and closes out the cedar
+// metadata if one was created with the completed_at timestamp and an
+// exit code.
+func (s *systemMetricsCollector) Close() error {
+	s.mu.Lock()
+	if s.closed {
+		s.mu.Unlock()
+		return errors.New("system metrics collector already cancelled or closed")
+	}
+	s.mu.Unlock()
+
+	if s.streamingCancel != nil {
+		s.streamingCancel()
+	}
+	s.close.Wait()
+	return s.catcher.Resolve()
+}
+
+func getSystemMetricsInfo(t *task.Task) *metrics.SystemMetricsOptions {
+	return &metrics.SystemMetricsOptions{
+		Project:     t.Project,
+		Version:     t.Version,
+		Variant:     t.BuildVariant,
+		TaskName:    t.DisplayName,
+		TaskId:      t.Id,
+		Execution:   int32(t.Execution),
+		Mainline:    !t.IsPatchRequest(),
+		Compression: metrics.CompressionTypeNone,
+		Schema:      metrics.SchemaTypeRawEvents,
+	}
+}
+
+func getDialOpts(ctx context.Context, c client.Communicator) (*timber.DialCedarOptions, error) {
```

Comment:
The client communicator holds a connection to the cedar grpc client, so one may already exist. We should probably refactor the code there and add a new func to the communicator interface that returns the grpc connection (creating a new one if it doesn't exist yet).